### PR TITLE
Use UnsafeCell for PgLwLock and PgAtomic

### LINKED ISSUE #1223: PgLwLock breaks postgres reinitialization after backend crash
Normally when a backend crashes with a segfault, postmaster will kill all other backends and then reinitialize everything. But this breaks when using `PgLwLock`, because shared memory initialization of the extension is executed again and then PgLwLock complains that it cannot attach.
```
2023-07-21 16:08:56.754 CEST [538131] LOG:  background worker "Some background worker" (PID 538137) was terminated by signal 11: Segmentation fault
2023-07-21 16:08:56.755 CEST [538131] LOG:  all server processes terminated; reinitializing
2023-07-21 16:08:56.924 CEST [538131] FATAL:  Can't attach, lock is not in an empty state: PgLwLockInner
```

### PR DESCRIPTION
**Pull request by JelteF**
As explained in detail in #1223, due to the Postgres restart-on-crash logic, `PgAtomic<T>::attach` and `PgLwLock<T>::attach` would be called again after a crash. Using OnceCell prevented the restarted extension from attaching new PgLwLocks, so Postgres failed to restart after a crash. This addresses that by using UnsafeCell instead of OnceCell and marking `attach` as unsafe.

Fixes #1223

### CODE REVIEW

File: pgrx/src/shmem.rs [hunk 1]
```diff
@@ -184,12 +187,11 @@ impl PgSharedMem {
...
             lock.attach(fv_shmem);
             pg_sys::LWLockRelease(addin_shmem_init_lock);
-        }
     }
 
     /// Must be run from the shared memory init hook, use for rust atomics behind `PgAtomic`
-    pub fn shmem_init_atomic<T: atomic_traits::Atomic + Default>(atomic: &PgAtomic<T>) {
-        unsafe {
+    /// SAFETY: Can only be called from Postgres its shared memory init hook
+    pub unsafe fn shmem_init_atomic<T: atomic_traits::Atomic + Default>(atomic: &PgAtomic<T>) {
```
Comment (workingjubilee) [inline pgrx/src/shmem.rs]: We can put this unsafe block back, and in so doing not have to reformat everything.
```suggestion
    /// # Safety
    /// Must only be called from Postgres its shared memory init hook
    pub unsafe fn shmem_init_atomic<T: atomic_traits::Atomic + Default>(atomic: &PgAtomic<T>) {
        unsafe {
```

File: pgrx/src/shmem.rs [hunk 2]
```diff
@@ -206,7 +208,6 @@ impl PgSharedMem {
             let atomic = T::default();
             std::ptr::copy(&atomic, fv_shmem, 1);
             pg_sys::LWLockRelease(addin_shmem_init_lock);
```
Comment (workingjubilee) [inline pgrx/src/shmem.rs]:
```suggestion
            pg_sys::LWLockRelease(addin_shmem_init_lock);
        }
```

File: pgrx/src/shmem.rs [hunk 3]
```diff
@@ -168,9 +171,9 @@ impl PgSharedMem {
     }
 
     /// Must be run from the shared memory init hook, use for types which are guarded by a `LWLock`
-    pub fn shmem_init_locked<T: Default + PGRXSharedMemory>(lock: &PgLwLock<T>) {
+    /// SAFETY: Can only be called from Postgres its shared memory init hook
+    pub unsafe fn shmem_init_locked<T: Default + PGRXSharedMemory>(lock: &PgLwLock<T>) {
```
Comment (workingjubilee) [inline pgrx/src/shmem.rs]: same here.

File: pgrx/src/shmem.rs
Comment (workingjubilee) [inline pgrx/src/shmem.rs:1]: Please add
```rust
#![deny(unsafe_op_in_unsafe_fn)]
```
if we're going to touch all the unsafe blocks in this file.

**Review by workingjubilee — CHANGES_REQUESTED** [review]

File: pgrx/src/shmem.rs
Comment (JelteF) [reply pgrx/src/shmem.rs:1]: All fixed now.

Comment (workingjubilee) [reply pgrx/src/shmem.rs:1]: nice.

File: pgrx/src/atomics.rs [hunk 4]
```diff
@@ -7,30 +7,32 @@
...
     }
 }
 
 impl<T> PgAtomic<T>
 where
     T: atomic_traits::Atomic + Default,
 {
-    pub fn attach(&self, value: *mut T) {
-        self.inner.set(value).expect("This PgAtomic is not empty, can't re-attach");
+    /// SAFETY: Can only be called from Postgres its shared memory init hook
```
Comment (workingjubilee) [inline pgrx/src/atomics.rs]: ...was slow to notice this typo. The extra "its"
```suggestion
    /// SAFETY: Must only be called from inside the Postgres shared memory init hook
```

**Review by workingjubilee — APPROVED** [review]
Ah, I had originally thought it was best to not change the safety of these fn unnecessarily, which is why I had dawdled to work out a slightly more sophisticated solution, but if this is what you would prefer, then that is fine. It's plausible you might be the heaviest user of this particular API currently, if my guess is correct.

By my evaluation, with the new safety constraint, this API is no more or less sound than it was.

One last nit, typo in the safety comment, and we'll be off to the races.

File: pgrx/src/atomics.rs (hunk 4 above)
Comment (JelteF) [reply pgrx/src/atomics.rs]: Applied everywhere

**Comment by JelteF** [comment]
> Ah, I had originally thought it was best to not change the safety of these fn unnecessarily, which is why I had dawdled to work out a slightly more sophisticated solution, but if this is what you would prefer, then that is fine.

Obviously safe functions are always nicer, but the nice thing is that the `pg_shmem_init!` macro is still marked as safe to call, as well as all the functions to lock/unlock the values. All the newly marked unsafe functions shouldn't really ever be called directly by people.

> It's plausible you might be the heaviest user of this particular API currently, if my guess is correct.

Possibly, and even I don't use it heavily.

> By my evaluation, with the new safety constraint, this API is no more or less sound than it was.

I'm not a huge expert in soundness of rust, but that was my assessment as well: I'm not 100% sure if it was sound before with all possible usages, but I don't think it is any worse than it is now. And now at least Postgres can restart after a crash.

**Comment by workingjubilee** [comment]
> Obviously safe functions are always nicer, but the nice thing is that the `pg_shmem_init!` macro is still marked as safe to call,

This is the thing that I am least confident of its correctness. But even if it does, that's because of a preexisting problem, so it shouldn't bar this PR. I am pretty sure it would at most be a one-time `unsafe` block with a trivial condition that amounts to "uh... don't... don't write really weird and complicated init hooks?" which is easy enough to satisfy.

> as well as all the functions to lock/unlock the values.

Yeah, these should always remain safe, and those are the ones that will actually be in most code.

**Pull request merged**